docs/CompilationExecution.md

```diff
@@ -0,0 +1,121 @@
+# Compilation and Execution
+
+This document contains an overview of the components involved in the process of compiling and executing a Flux query.
+
+In a nutshell, the process is separated in compilation and execution. The Flux compiler is responsible to turn the initial
+representation of a query into an Intermediate Representation (IR) that makes sense for the execution step.
+The Flux Virtual Machine ([FVM](VirtualMachine.md)) is in charge of interpreting the IR and execute it to output its results.
+In figure:
+
+![chain](compilation_execution.png)
+
+ - The initial representation of the query can be either a raw string or an already parsed Abstract Syntax Tree.
+ - The compilation process chains multiple compilation steps.
+ `Compiler`s turn a query representation in a language into another one, until they reach the required IR.
+ - The `Interpreter` takes the IR and interprets it.
+ Some operations are trivial ---e.g., `1 + 2` or `obj.b` or `arr[0]`--- and do not require any further step to be executed.
+ For data manipulation pipelines, instead, the `Intepreter` needs more.
+ - The `Interpreter` does not know how to execute data pipelines.
+ So, it takes their representation ---i.e., the `Spec`-- and optimizes it by applying subsequent _transformations_,
+ in order to delegates execution to the `Engine`.
+ Those transformations are logical and physical planning. We consider these transformations as further compilation
+ steps, in that they turn a query representation into another.
+ - Finally, the `Engine` executes the optimized representation and outputs the results.
+
+Wrapping up, the actors that play in this process are `Compiler`s ---that turn a representation in some language into another one---
+and `Executor`s ---that execute a representation and produce results.
+Both the `Interpreter` and the `Engine` are `Executor`s:
+ - The `Engine` actually executes data manipulation pipelines and produces results;
```

Comment:
From a previous comment from someone else, I like runtime as a better name. I do like that the engine/runtime do not have an interface description. To my mind, when we register functions, we'll register AST to op spec transformations with the interpreter and the engine/runtime will be an abstract separation.

I think this will also make things like the repl possible and use the same proxy behavior we have currently. So the repl will know how to transform AST to op spec and then will use an interface to send that spec to a server. We'll expose as a repl endpoint some version constraint API and generally require the repl to be the same version as the server to avoid errors similar to how postgres' client has requirements for the versioning.

Reply:
Yeah, that was me actually 😅 

Would you like to change `Engine` into `Runtime` 🤔 ?

To me `Runtime` sounds more passive than `Engine` which sounds like something that does things. I don't know, only matter of taste. I am not a native speaker too 😅 

I proposed `Runtime` for the object that gets things registered and provides them to `Executors`.